### PR TITLE
Prop Validation

### LINKED ISSUE #750: Propulsion Firings not Happening in HOOTL
For some reason the prop system ends up handling a fault. See #746 for details.

### LINKED ISSUE #761: [FSW] Rebind Statefields OrbitController x PropController
This is a classic example of why we need integrated testing or perhaps even a full code review.
We passed all our unit tests (only 1 control task), but we wouldn't have caught this if we hadn't done an integrated test with multiple control tasks! :D 

The sched valve fields in OrbitController are not bound PropController

image
image

### PR DESCRIPTION
**Pull request by govindchari**
# Prop Fixes

Fixes #761 #750 

### Summary of changes
- Rebound statefields to avoid valve schedule duplication
- To fix #750 you must ws the Tank2 pressure above the threshold to transition from pressurizing into await firing
- Added a Tank2 pressure for nearfield (currently it is arbitrarily set at 15psi)
- Added a second set of regression constants to convert the desired impulse into thruster on times

TODO:
Choose good value for Nearfield Tank2 pressure
Compute the regression constants based on theory (ideally would perform empirical tests)

### Testing
In HOOTL a firing does occur as given by
Screen Shot 2021-05-11 at 11 00 08 PM
Also the threshold is working
Screen Shot 2021-05-11 at 11 00 25 PM

### Constants
Two tracked constants are added: valve_time_lin_reg_intercept_near and valve_time_lin_reg_slope_near

### CODE REVIEW

File: src/fsw/FCCode/PropController.cpp [hunk 1]
```diff
@@ -97,13 +97,27 @@ PropState_Venting PropController::state_venting;
 PropState_HandlingFault PropController::state_handling_fault;
 PropState_Manual PropController::state_manual;
 
+void init(){
+    //sched_valve1_fp = FIND_WRITABLE_FIELD(unsigned int, "orbit.control.valve1");
+    //sched_valve2_fp = FIND_WRITABLE_FIELD(unsigned int, "orbit.control.valve2");
+    //sched_valve3_fp = FIND_WRITABLE_FIELD(unsigned int, "orbit.control.valve3");
+    //sched_valve4_fp = FIND_WRITABLE_FIELD(unsigned int, "orbit.control.valve4");
+}
+
```
Comment (fatimayousuf) [inline src/fsw/FCCode/PropController.cpp:106]: I think you can just get rid of this now?

File: src/fsw/FCCode/PropController.hpp [hunk 2]
```diff
@@ -223,6 +225,8 @@ class PropState
     //      that means we are not changing states this cycle.
     virtual prop_state_t evaluate() = 0;
 
+    void init();
+
```
Comment (fatimayousuf) [inline src/fsw/FCCode/PropController.hpp:229]: delete this too

File: src/fsw/FCCode/OrbitController.cpp [hunk 3]
```diff
@@ -222,13 +224,19 @@ lin::Vector3d OrbitController::calculate_impulse(double t, const lin::Vector3d &
...
 }
 
-unsigned int OrbitController::impulse_to_time(double impulse) {
+unsigned int OrbitController::impulse_to_time(double impulse, unsigned char state) {
+    double valve_time_lin_reg_slope = valve_time_lin_reg_slope_far;
+    double valve_time_lin_reg_intercept = valve_time_lin_reg_intercept_far;
+    if (state==static_cast<unsigned char>(rel_orbit_state_t::estimating)) {
+        valve_time_lin_reg_slope = valve_time_lin_reg_slope_near;
+        valve_time_lin_reg_intercept = valve_time_lin_reg_intercept_near;
+    }
```
Comment (fatimayousuf) [inline src/fsw/FCCode/OrbitController.cpp:233]: Can you move this calculation to the execute() function like you did in Prop Controller, perhaps before when schedule_valves() is called? Then you don't need to pass the state to schedule_valves() or impulse_to_time().

File: src/fsw/FCCode/PropController.cpp [hunk 4]
```diff
@@ -675,10 +689,10 @@ void manual_eval(WritableStateField<unsigned int> &sched,
...
 {
-    manual_eval(controller->sched_valve1_f, Tank1, 0);
-    manual_eval(controller->sched_valve2_f, Tank1, 1);
-    manual_eval(controller->sched_valve3_f, Tank1, 2);
-    manual_eval(controller->sched_valve4_f, Tank1, 3);
+    manual_eval(*(controller->sched_valve1_fp), Tank1, 0);
+    manual_eval(*(controller->sched_valve2_fp), Tank1, 1);
+    manual_eval(*(controller->sched_valve3_fp), Tank1, 2);
+    manual_eval(*(controller->sched_valve4_fp), Tank1, 3);
     manual_eval(controller->sched_intertank1_f, Tank2, 0);
```
Comment (fatimayousuf) [inline src/fsw/FCCode/PropController.cpp:696]: just wondering but should this be `manual_eval(controller->sched_intertank1_f, Tank1, 0);`?

**Review by fatimayousuf — CHANGES_REQUESTED** [review]
Looks great!

File: ptest/cases/mission/near_field.py [hunk 5]
```diff
@@ -1,18 +1,22 @@
...
 
 class DualSatNearFieldCase(DualSatCase, PSimCase):
 
     def __init__(self, *args, **kwargs):
         super(DualSatNearFieldCase, self).__init__(*args, **kwargs)
 
         self.psim_configs += ["truth/near_field"]
+        # self.psim_config_overrides = {"truth.leader.attitude.w": lin.Vector3([0,0,0]), "truth.follower.attitude.w": lin.Vector3([0,0,0])}
+
+        # self.psim_config_overrides = {"truth.leader.pos": lin.Vector3[0,0,0,0], "truth.leader.vel"}
```
Comment (tanishqaggarwal) [inline ptest/cases/mission/near_field.py:14]:
```suggestion
```

File: ptest/cases/mission/near_field.py [hunk 6]
```diff
@@ -24,7 +28,8 @@ def post_boot(self):
 
         self.flight_controller_leader.write_state("pan.sat_designation", Enums.sat_designations["leader"])
         self.flight_controller_follower.write_state("pan.sat_designation", Enums.sat_designations["follower"])
-
+        # self.flight_controller_follower.ws("prop.threshold_firing")
+        # self.flight_controller_leader.write_state("dcdc.SpikeDock_cmd
```
Comment (tanishqaggarwal) [inline ptest/cases/mission/near_field.py:32]:
```suggestion
```

File: ptest/cases/mission/utils.py [hunk 7]
```diff
@@ -31,11 +32,24 @@
...
-#    "piksi.time"
-#    "adcs_monitor.ssa_vec",
-#    "adcs_monitor.mag1_vec",
-#    "adcs_monitor.gyr_vec",
-#    "adcs_monitor.ssa_mode",
+    #    "piksi.time"
+    #    "adcs_monitor.ssa_vec",
+    #    "adcs_monitor.mag1_vec",
+    #    "adcs_monitor.gyr_vec",
+    #    "adcs_monitor.ssa_mode",
```
Comment (tanishqaggarwal) [inline ptest/cases/mission/utils.py:39]:
```suggestion
```

File: src/flow_data.cpp [hunk 8]
```diff
@@ -31,7 +31,7 @@ const std::vector<DownlinkProducer::FlowData> PAN::flow_data = {
 {24, false, {"adcs_monitor.ssa_voltage10", "adcs_monitor.ssa_voltage11", "adcs_monitor.ssa_voltage12", "adcs_monitor.ssa_voltage13", "adcs_monitor.ssa_voltage14", "adcs_monitor.ssa_voltage15", "adcs_monitor.ssa_voltage16", "adcs_monitor.ssa_voltage17", "adcs_monitor.ssa_voltage18", "adcs_monitor.ssa_voltage19"}},
 {25, true, {"docksys.docked"}},
 {26, true, {"orbit.control.valve1", "orbit.control.valve2", "orbit.control.valve3", "orbit.control.valve4", "orbit.control.J_ecef", "orbit.control.alpha"}},
-{27, false, {"prop.cycles_until_firing", "prop.sched_valve1", "prop.sched_valve2", "prop.sched_valve3", "prop.sched_valve4", "prop.sched_intertank1", "prop.sched_intertank2"}},
+{27, false, {"prop.cycles_until_firing", "prop.sched_intertank1", "prop.sched_intertank2"}},
```
Comment (tanishqaggarwal) [inline src/flow_data.cpp:34]: Why remove these values from flow data?

File: src/fsw/FCCode/OrbitController.hpp [hunk 9]
```diff
@@ -47,13 +51,13 @@ class OrbitController : public TimedControlTask<void>
     /*
      * Convert the impulse of a thruster to the time the valve should be open in ms
      */
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/OrbitController.hpp:53]: Add docstring for function arguments?

File: src/fsw/FCCode/OrbitController.hpp [hunk 10]
```diff
@@ -47,13 +51,13 @@ class OrbitController : public TimedControlTask<void>
...
-    unsigned int impulse_to_time(double impulse);
+    unsigned int impulse_to_time(double impulse, unsigned char state);
 
     /**
      * Calculates the time each valve should open to deliver a given impulse. 
      * The impulse must be in the body frame of the satellite
      */
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/OrbitController.hpp:59]: Add docstring for function arguments?

File: src/fsw/FCCode/PropController.cpp [hunk 11]
```diff
@@ -104,6 +104,13 @@ void PropController::execute()
     if (cycles_until_firing.get() > 0)
         cycles_until_firing.set(cycles_until_firing.get() - 1);
 
+    // If in nearfield Tank 2 pressure should be lower than in far-field
+    unsigned char rel_orbit_state=rel_orbit_valid_fp->get();
+    threshold_firing_pressure.set(threshold_firing_pressure_far_ic);
+    if (rel_orbit_state == static_cast<unsigned char>(rel_orbit_state_t::estimating)) {
+        threshold_firing_pressure.set(threshold_firing_pressure_near_ic);
+    }
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/PropController.cpp:112]: Software design-wise, I feel like this logic should exist in orbit controller; prop controller shouldn't care about orbital state

File: src/fsw/FCCode/PropController.cpp [hunk 12]
```diff
@@ -11,12 +13,14 @@
...
       prop_state_f("prop.state", Serializer<unsigned int>(9)),
       cycles_until_firing("prop.cycles_until_firing", Serializer<unsigned int>(orbit_ccno)),
-      sched_valve1_f("prop.sched_valve1", Serializer<unsigned int>(999)),
-      sched_valve2_f("prop.sched_valve2", Serializer<unsigned int>(999)),
-      sched_valve3_f("prop.sched_valve3", Serializer<unsigned int>(999)),
-      sched_valve4_f("prop.sched_valve4", Serializer<unsigned int>(999)),
+      sched_valve1_fp(FIND_WRITABLE_FIELD(unsigned int, orbit.control.valve1)),
+      sched_valve2_fp(FIND_WRITABLE_FIELD(unsigned int, orbit.control.valve2)),
+      sched_valve3_fp(FIND_WRITABLE_FIELD(unsigned int, orbit.control.valve3)),
+      sched_valve4_fp(FIND_WRITABLE_FIELD(unsigned int, orbit.control.valve4)),
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/PropController.cpp:23]: I see that the valve schedule fields have moved to orbit controller; but are we still able to test the prop system manually by setting schedules?

File: src/fsw/FCCode/PropController.cpp (hunk 11 above)
Comment (shihaocao) [inline src/fsw/FCCode/PropController.cpp:112]: We need to actually have two new statefields that are both Writable

`threshold_firing_that_is_actually_used` set to `thresh_1` in nominal conditions, but `thresh_2` when we are in near field

File: test/test_fsw_prop_faults/test_prop_faults.cpp [hunk 13]
```diff
@@ -153,7 +153,7 @@ void test_underpressured_unsignal_response()
     }
     // If the ground decides to unsignal and lower the pressure threshold
     unsignal_fault(pressurize_fail_fault_f);
-    tf.pc->threshold_firing_pressure.set(14.0); // default fake pressure is 14.6 psi
```
Comment (shihaocao) [inline test/test_fsw_prop_faults/test_prop_faults.cpp:156]: This brings down the tunable field that we would use on the ground to fake it being above the threshold pressure

File: src/flow_data.cpp (hunk 8 above)
Comment (fatimayousuf) [reply src/flow_data.cpp:34]: #761

Comment (tanishqaggarwal) [reply src/flow_data.cpp:34]: Lol I'm blind, I see the orbit control valve state fields now

**Review by fatimayousuf — APPROVED** [review]

**Pull request merged**